fennel/internal_lib/to_proto/source_code.py

```diff
@@ -101,15 +100,78 @@ def get_featureset_gen_code(
     return _remove_empty_lines(source_code)
 
 
-def remove_source_and_sink_decorator(text):
-    # Define the regular expression pattern for the @source decorator block
-    pattern = r"^\s*@source\((?:.|\n)*?\)\s*$"
-    # Remove @source decorator using the regex pattern with re.MULTILINE and re.DOTALL flags
-    result = re.sub(pattern, "", text, flags=re.MULTILINE | re.DOTALL)
-    pattern = r"^\s*@sink\((?:.|\n)*?\)\s*$"
-    # Remove @source decorator using the regex pattern with re.MULTILINE and re.DOTALL flags
-    result = re.sub(pattern, "", result, flags=re.MULTILINE | re.DOTALL)
-    return result.strip()
+def get_text_between_ast_pointers(
+    text: str,
+    lineno: int,
+    col_offset: int,
+    end_lineno: int,
+    end_col_offset: int,
+    is_decorator: bool = False,
+):
+    # Split the text into lines
+    lines = text.splitlines()
+
+    if is_decorator:
+        # for adding @
+        col_offset = col_offset - 1
+
+    output = []
+    if lineno == end_lineno:
+        return lines[lineno - 1][col_offset:end_col_offset]
+    for i in range(lineno - 1, end_lineno):
+        if i == lineno - 1:
+            output.append(lines[i][col_offset:])
+        elif i == end_lineno - 1:
+            output.append(lines[i][:end_col_offset])
+        else:
+            output.append(lines[i])
+    return "\n".join(output)
+
+
+class DecoratorRemover:
+    def __init__(self, text: str, decorator_list: List[str]):
+        self.text = text
+        self.decorator_list = decorator_list
+        self.decorator_lines: List[str] = []
+
+    def get_decorators(self, node):
+        decorator_list = [
+            decorator
+            for decorator in node.decorator_list
+            if isinstance(decorator, ast.Call)
+            and hasattr(decorator.func, "id")
+            and getattr(decorator.func, "id") in self.decorator_list
+        ]
+        for decorator in decorator_list:
+            self.decorator_lines.append(
+                get_text_between_ast_pointers(
+                    self.text,
+                    decorator.lineno,
+                    decorator.col_offset,
+                    decorator.end_lineno,
+                    decorator.end_col_offset,
+                    True,
+                )
+            )
+        return node
+
+    def get_filtered_text(self) -> str:
+        generator = ast.walk(ast.parse(self.text))
+        next(
+            (
+                self.get_decorators(node)
+                for node in generator
+                if isinstance(node, (ast.ClassDef, ast.FunctionDef))
+            ),
+            None,
+        )
+        for line in self.decorator_lines:
+            self.text = self.text.replace(line, "")
+        return self.text.strip()
+
+
+def remove_source_and_sink_decorator(text: str):
```

Comment:
Makes source and sink a parameter, so that it can be used by anyone else in the future. 

Can you add a test suite for this function, for different decorators say meta etc

Reply:
added.